Unexpected exception "TypeGaurd: Segement Must exist Must exist"

### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.


### What Happened

Executed `sqlfluff lint` but it gave an exception

```
➜ sqlfluff lint file.sql --dialect mysql
CRITICAL   [L049] Applying rule L049 threw an Exception: TypeGaurd: Segement Must exist Must exist                                                                                                                                        
Traceback (most recent call last):
  File "/home/jasper/.local/lib/python3.8/site-packages/sqlfluff/core/rules/base.py", line 644, in crawl
    res = self._eval(context=context)
  File "/home/jasper/.local/lib/python3.8/site-packages/sqlfluff/rules/L049.py", line 82, in _eval
    assert sub_seg, "TypeGaurd: Segement Must exist Must exist"
AssertionError: TypeGaurd: Segement Must exist Must exist
== [file.sql] FAIL                                                                                                                                
L:   1 | P:   1 | L049 | Unexpected exception: TypeGaurd: Segement Must exist
                       | Must exist; Could you open an issue at
                       | https://github.com/sqlfluff/sqlfluff/issues ? You can
                       | ignore this exception for now, by adding '-- noqa: L049'
                       | at the end of line 1
L:   1 | P:  11 |  PRS | Line 1, Position 11: Found unparsable section: '(SELECT
                       | id FROM table WHERE field = TRUE...'
All Finished 📜 🎉!
```

### Expected Behaviour

No exception but maybe errors or warnings on the line when needed

### Observed Behaviour

The program died where it should not have.

### How to reproduce

My file contained multiple lines, but died on the first one:

```sql
SET @id = (SELECT id FROM table WHERE field = TRUE LIMIT 1);

UPDATE other_table SET foreign_key = @id WHERE foreign_key IS NULL;
```

### Dialect

mysql

### Version

➜  sqlfluff --version
sqlfluff, version 0.13.2
➜  python --version
Python 3.8.10


### Configuration

the default config

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

--- a/src/sqlfluff/dialects/dialect_mysql.py
+++ b/src/sqlfluff/dialects/dialect_mysql.py
@@ -1084,6 +1084,7 @@ class SetAssignmentStatementSegment(BaseSegment):
                     Ref("LocalVariableNameSegment"),
                     Ref("FunctionSegment"),
                     Ref("ArithmeticBinaryOperatorGrammar"),
+                    Ref("ExpressionSegment"),
                 ),
             ),
         ),

--- a/test/fixtures/dialects/mysql/repeat_label.yml
+++ b/test/fixtures/dialects/mysql/repeat_label.yml
@@ -3,7 +3,7 @@
 # computed by SQLFluff when running the tests. Please run
 # `python test/generate_parse_fixture_yml.py`  to generate them after adding or
 # altering SQL files.
-_hash: b6d7da809dc5940192ca1a1766435ebadffeb15165190b09fe54dee358f939ad
+_hash: 1ec4cd7ff211f8acd0d2bf094c7fd45cf49c64e79da0292d7d8619d2a64189f1
 file:
 - statement:
     repeat_statement:
@@ -12,13 +12,15 @@ file:
       keyword: repeat
       statement:
         set_statement:
-        - keyword: set
-        - variable: '@a'
-        - comparison_operator:
+          keyword: set
+          variable: '@a'
+          comparison_operator:
             raw_comparison_operator: '='
-        - variable: '@a'
-        - binary_operator: +
-        - variable: '1'
+          expression:
+            column_reference:
+              variable: '@a'
+            binary_operator: +
+            literal: '1'
 - statement_terminator: ;
 - statement:
     repeat_statement:

--- a/test/fixtures/dialects/mysql/repeat_multiple_statements.yml
+++ b/test/fixtures/dialects/mysql/repeat_multiple_statements.yml
@@ -3,20 +3,22 @@
 # computed by SQLFluff when running the tests. Please run
 # `python test/generate_parse_fixture_yml.py`  to generate them after adding or
 # altering SQL files.
-_hash: 2a4efc1bff55db536def1f2ac007ece8c5ebaa1367c2cdcc782ba6b41867216f
+_hash: 404e91a1d133223c6a6b3d71c8b98914f5fd0ae71c024dbf38931ef12024652c
 file:
 - statement:
     repeat_statement:
       keyword: repeat
       statement:
         set_statement:
-        - keyword: set
-        - variable: '@a'
-        - comparison_operator:
+          keyword: set
+          variable: '@a'
+          comparison_operator:
             raw_comparison_operator: '='
-        - variable: '@a'
-        - binary_operator: +
-        - variable: '1'
+          expression:
+            column_reference:
+              variable: '@a'
+            binary_operator: +
+            literal: '1'
 - statement_terminator: ;
 - statement:
     select_statement:

--- a/test/fixtures/dialects/mysql/repeat_no_label.yml
+++ b/test/fixtures/dialects/mysql/repeat_no_label.yml
@@ -3,20 +3,22 @@
 # computed by SQLFluff when running the tests. Please run
 # `python test/generate_parse_fixture_yml.py`  to generate them after adding or
 # altering SQL files.
-_hash: 92cf623eb79841e26052557d8edebb1e12628046ad6f4938bcc2f1dc5c9528f8
+_hash: 3b4d4e126f802dff131b90d6730fae8d7cf5524a89a2e0ae73b4aaee341c1459
 file:
 - statement:
     repeat_statement:
       keyword: repeat
       statement:
         set_statement:
-        - keyword: set
-        - variable: '@a'
-        - comparison_operator:
+          keyword: set
+          variable: '@a'
+          comparison_operator:
             raw_comparison_operator: '='
-        - variable: '@a'
-        - binary_operator: +
-        - variable: '1'
+          expression:
+            column_reference:
+              variable: '@a'
+            binary_operator: +
+            literal: '1'
 - statement_terminator: ;
 - statement:
     repeat_statement:

--- a/test/fixtures/dialects/mysql/set_session_variable_expression.sql
+++ b/test/fixtures/dialects/mysql/set_session_variable_expression.sql
@@ -1,1 +1,3 @@
 set @abc = 1 + 2;
+set @abc = (select 1);
+SET @id = (SELECT id FROM table1 WHERE field = TRUE LIMIT 1);

--- a/test/fixtures/dialects/mysql/set_session_variable_expression.yml
+++ b/test/fixtures/dialects/mysql/set_session_variable_expression.yml
@@ -3,15 +3,69 @@
 # computed by SQLFluff when running the tests. Please run
 # `python test/generate_parse_fixture_yml.py`  to generate them after adding or
 # altering SQL files.
-_hash: cd75ed2c08475dec11a6995c52221ce6beb33fd0c9c4d9f2ddb24c2c3861ed73
+_hash: 082a96fc948a0732e51dea89daf65100cf3d033e5b684c7337a690fe365a626c
 file:
-  statement:
+- statement:
     set_statement:
-    - keyword: set
-    - variable: '@abc'
-    - comparison_operator:
+      keyword: set
+      variable: '@abc'
+      comparison_operator:
         raw_comparison_operator: '='
-    - variable: '1'
-    - binary_operator: +
-    - variable: '2'
-  statement_terminator: ;
+      expression:
+      - literal: '1'
+      - binary_operator: +
+      - literal: '2'
+- statement_terminator: ;
+- statement:
+    set_statement:
+      keyword: set
+      variable: '@abc'
+      comparison_operator:
+        raw_comparison_operator: '='
+      expression:
+        bracketed:
+          start_bracket: (
+          expression:
+            select_statement:
+              select_clause:
+                keyword: select
+                select_clause_element:
+                  literal: '1'
+          end_bracket: )
+- statement_terminator: ;
+- statement:
+    set_statement:
+      keyword: SET
+      variable: '@id'
+      comparison_operator:
+        raw_comparison_operator: '='
+      expression:
+        bracketed:
+          start_bracket: (
+          expression:
+            select_statement:
+              select_clause:
+                keyword: SELECT
+                select_clause_element:
+                  column_reference:
+                    identifier: id
+              from_clause:
+                keyword: FROM
+                from_expression:
+                  from_expression_element:
+                    table_expression:
+                      table_reference:
+                        identifier: table1
+              where_clause:
+                keyword: WHERE
+                expression:
+                  column_reference:
+                    identifier: field
+                  comparison_operator:
+                    raw_comparison_operator: '='
+                  literal: 'TRUE'
+              limit_clause:
+                keyword: LIMIT
+                literal: '1'
+          end_bracket: )
+- statement_terminator: ;

--- a/test/fixtures/dialects/mysql/while_label.yml
+++ b/test/fixtures/dialects/mysql/while_label.yml
@@ -3,7 +3,7 @@
 # computed by SQLFluff when running the tests. Please run
 # `python test/generate_parse_fixture_yml.py`  to generate them after adding or
 # altering SQL files.
-_hash: eac8745f9cb4e0218ca54417e1728db6b88cc040b2e8fad34f5dc375d4e687bd
+_hash: e6b9836bf30f693cb2139d215af5afa614e5f43e2072c0f62799e8d42a5ca12b
 file:
 - statement:
     while_statement:
@@ -21,13 +21,15 @@ file:
     - keyword: do
     - statement:
         set_statement:
-        - keyword: set
-        - variable: _cnt
-        - comparison_operator:
+          keyword: set
+          variable: _cnt
+          comparison_operator:
             raw_comparison_operator: '='
-        - variable: _cnt
-        - binary_operator: +
-        - variable: '1'
+          expression:
+            column_reference:
+              identifier: _cnt
+            binary_operator: +
+            literal: '1'
 - statement_terminator: ;
 - statement:
     while_statement:

--- a/test/fixtures/dialects/mysql/while_no_label.yml
+++ b/test/fixtures/dialects/mysql/while_no_label.yml
@@ -3,7 +3,7 @@
 # computed by SQLFluff when running the tests. Please run
 # `python test/generate_parse_fixture_yml.py`  to generate them after adding or
 # altering SQL files.
-_hash: 8c71de0ce64e132e00750b7eebe5dce7d20b651e3fc319bb1d387034b77d8649
+_hash: 1cd9aef4dbdd3fb8cc654605b500ada371591855f7dea7c93e33722d00ad4181
 file:
 - statement:
     while_statement:
@@ -19,13 +19,15 @@ file:
     - keyword: do
     - statement:
         set_statement:
-        - keyword: set
-        - variable: _cnt
-        - comparison_operator:
+          keyword: set
+          variable: _cnt
+          comparison_operator:
             raw_comparison_operator: '='
-        - variable: _cnt
-        - binary_operator: +
-        - variable: '1'
+          expression:
+            column_reference:
+              identifier: _cnt
+            binary_operator: +
+            literal: '1'
 - statement_terminator: ;
 - statement:
     while_statement:

